asyncscope.md

```diff
@@ -394,15 +451,34 @@ This is considered boilerplate and not helpful, as the sopped scenarios should b
 
 ## Stop handling
 
-The paper requires that if the caller requests stop to an `async_scope` object, then the work that is currently being executed is not cancelled (unless the work explicitly interacts with the stop source of the `async_scope` object).
+The paper requires that if the caller requests stop to an `async_scope` object, then the work that is currently being executed is not stopped (unless the work explicitly interacts with the `stop_source` of the `async_scope` object).
+
+### `request_stop` on the `async_scope` is not forwarded
+
+A `stop_callback` is not a destructor, it is a signal requesting running work to stop. Calling `stop_source::request_stop()` from the destructor is a bug, because that would use a **sync** scope exit to implicitly request that **async** senders stop. In the case of an `async_scope`, this bug is particularly egregious, since the sync scope exit may not own all the code for all the senders that were nested and spawned within the `async_scope`.
+
```

Comment:
I'm afraid I don't understand the problem. Can you please explain what would be the problem here?

I was just thinking yesterday, that maybe we should make `async_scope` behave like `jthread` and ensure that the destructor waits for all the work. Something like the following:
* destructor calls `request_stop()`
* destructor creates a semaphore
* destructor registers a sender on `on_empty()` that will signal the semaphore
* destructor waits for the semaphore to comple

This strategy is not perfect. One can imagine cases in which senders that are passed to `spawn()` will not make progress until called from exterior, and cannot be cancelled. This can lead to a deadlock. But this is a bug in the sender passed in to `spawn()`, not a problematic case for `async_scope`. Similar problems can be created with `jthread` too.

Reply:
>  we should make async_scope behave like jthread and ensure that the destructor waits for all the work.

This is suboptimal behaviour though for a lot of uses. It just makes systems prone to deadlock instead of termination, which isn't obviously an improvement. I have never found the arguments in favour of jthread convincing for this reason.

> `request_stop` on the `async_scope` is not forwarded

There are two different types of stop we need here. One of them we absolutely need: to be able to stop the scope accepting more work. That allows us to shut down cleanly because we tell the scope to reject work, then we wait for (in some fashion) the on_empty sender and we know we have completed the work.

It's optional whether we on top of that request that work that has been queued be cancelled.

We can build the latter on top of the former by injecting the stop_token when nest is called. We can't do the reverse. We can build that into the scope by propagating the stop source and allowing that stop source to be independently cancelled, or we can move it out of band entirely by passing the stop source around separately.

Reply:
I added this to a new Q/A section

Reply:
> I say that request_stop() should not be called from a destructor because a sync destructor should not be signaling that async operations should be discarded without waiting for them to complete.

There's still the nuance there though. At some point we do need to stop accepting new work. Is it ok to do that in a destructor, as distinct from discarding already queued work and waiting for the stop signal?

Although maybe what you are saying is that because all of this is async, apart from `spawn`, we always need to wait asynchronously even for an immediate discard signal, which means waiting for someone to `connect` and `start` the work, and none of that should happen in a destructor.

Reply:
Yes, having no way to stop adding new work makes the clean shutdown case more complicated to explain, but better motivates getting the structure correct. the call to start the empty_sender must happen-after all calls to nest(), spawn() and spawn_future(). This means that the operations adding to the scope must complete before starting the empty_sender. This is how structured concurrency should be used anyway. 

A separate mechanism to lock the async_scope so that no new work can be added leaves design issues about how to 'fail' nest(), spawn() and spawn_future(). Defining these failures would guarantee that usage would handle the errors rather than structuring the work..